lib/rucio/core/replica.py

```diff
@@ -2745,138 +2991,153 @@ def list_dataset_replicas(scope, name, deep=False, *, session: "Session"):
+            sub_query_stmt.c.rse_id,
+            sub_query_stmt.c.rse
+        ).subquery()
+
+        full_query_stmt = select(
+            group_query_stmt.c.dataset_scope.label('scope'),
+            group_query_stmt.c.dataset_name.label('name'),
+            group_query_stmt.c.rse_id,
+            group_query_stmt.c.rse,
+            func.sum(group_query_stmt.c.file_bytes).label('available_bytes'),
+            func.count().label('available_length'),
+            func.min(group_query_stmt.c.created_at).label('created_at'),
+            func.max(group_query_stmt.c.updated_at).label('updated_at'),
+            func.max(group_query_stmt.c.accessed_at).label('accessed_at')
+        ).group_by(
+            group_query_stmt.c.dataset_scope,
+            group_query_stmt.c.dataset_name,
+            group_query_stmt.c.rse_id,
+            group_query_stmt.c.rse
+        )
+
+        sub_query_stmt = select(
+            models.DataIdentifierAssociation.scope,
+            models.DataIdentifierAssociation.name,
+            models.RSEFileAssociation.rse_id,
+            func.sum(models.RSEFileAssociation.bytes).label("available_bytes"),
+            func.count().label("available_length"),
+            func.min(models.RSEFileAssociation.created_at).label("created_at"),
+            func.max(models.RSEFileAssociation.updated_at).label("updated_at"),
+            func.max(models.RSEFileAssociation.accessed_at).label("accessed_at")
+        ).with_hint(
+            models.DataIdentifierAssociation,
+            'INDEX_RS_ASC(CONTENTS CONTENTS_PK) INDEX_RS_ASC(REPLICAS REPLICAS_PK) NO_INDEX_FFS(CONTENTS CONTENTS_PK)', 'oracle'
+        ).where(
+            and_(models.DataIdentifierAssociation.child_scope == models.RSEFileAssociation.scope,
+                 models.DataIdentifierAssociation.child_name == models.RSEFileAssociation.name,
+                 models.DataIdentifierAssociation.scope == scope,
+                 models.DataIdentifierAssociation.name == name,
+                 models.RSEFileAssociation.state == ReplicaState.AVAILABLE)
+        ).group_by(
+            models.DataIdentifierAssociation.scope,
+            models.DataIdentifierAssociation.name,
+            models.RSEFileAssociation.rse_id
+        ).subquery()
+
+        stmt = select(
+            sub_query_stmt.c.scope,
+            sub_query_stmt.c.name,
+            sub_query_stmt.c.rse_id,
+            models.RSE.rse,
+            sub_query_stmt.c.available_bytes,
+            sub_query_stmt.c.available_length,
+            sub_query_stmt.c.created_at,
+            sub_query_stmt.c.updated_at,
+            sub_query_stmt.c.accessed_at
+        ).where(
+            and_(sub_query_stmt.c.rse_id == models.RSE.id,
+                 models.RSE.deleted == false())
+        )
+
+        # Join everything together
+        final_stmt = stmt.union_all(full_query_stmt)
+        for row in session.execute(final_stmt).all():
             replica = row._asdict()
```

Comment:
I think the more efficent version of the query.all() -> row._asdict() is to use [mappings](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.mappings)

Reply:
I am not sure we can do this, `mappings` returns an object (RowMapping) that does not implement `__setitem__`

```
"__setitem__" method not defined on type "RowMapping”
```